### ISSUE TITLE
Strange error: ReqlClientError: MALFORMED PROTOBUF (missing field `type`)

### ISSUE THREAD

**Issue by thelinuxlich** [neumino/rethinkdbdash#153]
```
 File "/app/releases/20150916042638/node_modules/rethinkdbdash/lib/connection.js", line 196, in Connection._processResponse
    currentReject(new Err.ReqlClientError(helper.makeAtom(response), self.metadata[token].query, response));
  File "/app/releases/20150916042638/node_modules/rethinkdbdash/lib/connection.js", line 150, in Socket.<anonymous>
    self._processResponse(response, token);
  File "events.js", line 77, in emitOne
  File "events.js", line 169, in Socket.emit
  File "_stream_readable.js", line 146, in readableAddChunk
  File "_stream_readable.js", line 110, in Socket.Readable.push
  File "net.js", line 523, in TCP.onread
```

**Comment by thelinuxlich** [comment]
The query: 

```
r.table("deleted_sessions").between([true, r.minval], [true, r.now().inTimezone("-03:00").sub(3600)], {
    index: "last_pass_created_at"
}).delete({
    durability: "soft",
    returnChanges: true
})("changes")("old_val").default([]).forEach(function(var_159) {
    return r.branch(var_159.ne(null), r.table("sessions").get(var_159("id")).delete({
        durability: "soft"
    }).do(function() {
        return r.expr(["events", "errors", "forms", "purchases", "mailings"]).concatMap(function(var_160) {
            return r.table(var_160).getAll(var_159("id"), {
                index: "session_id"
            }).map(function(var_161) {
                return {
                    batch: [{
                        table: var_160,
                        data: var_161.merge(r.branch(var_161.hasFields("updated_at"), {
                            visitor_id: var_159("new_visitor_id"),
                            session_id: var_159("new_session_id"),
                            created_at: var_161("created_at").toISO8601(),
                            ended_at: var_161("updated_at").toISO8601()
                        }, {
                            visitor_id: var_159("new_visitor_id"),
                            session_id: var_159("new_session_id"),
                            created_at: var_161("created_at").toISO8601()
                        }))
                    }]
                }
            })
        })
    }).do(function(var_162) {
        return r.expr(["traits", "pageviews", "subscriptions", "impressions", "session_goals", "visitor_goals"])
            .forEach(function(var_163) {
                return r.table(var_163).getAll(var_159("id"), {
                    index: "session_id"
                }).delete({
                    durability: "soft",
                    returnChanges: true
                })("changes")("old_val").default([]).map(function(var_164) {
                    return {
                        batch: [{
                            table: var_163,
                            data: var_164.merge({
                                id: r.literal(),
                                created_at: var_164("created_at").toISO8601(),
                                visitor_id: var_159("new_visitor_id"),
                                session_id: var_159("new_session_id")
                            }).merge(r.branch(var_163.eq("subscriptions"), {
                                email: var_159("new_email")
                            }, {})).merge(r.branch(var_164.hasFields("ended_at"), {
                                ended_at: var_164("ended_at").toISO8601()
                            }, r.branch(var_164.hasFields("updated_at"), {
                                updated_at: var_164("updated_at").toISO8601()
                            }, {})))
                        }]
                    }
                }).coerceTo("array").add(var_162)
            })
    }), [])
})("batch").distinct().default([])
```

**Comment by neumino** [comment]
This query works fine for me. Do you dynamically generate this query?

**Comment by thelinuxlich** [comment]
Yeah

**Comment by thelinuxlich** [comment]
I've changed to the official driver, without changing any logic and the error disappeared

**Comment by thelinuxlich** [comment]
The failing script:

```javascript
"use strict";

var moment = require("moment-timezone"),
    nano = require("nanomsg"),
    mq = nano.socket("push", {
        tcpnodelay: true
    }),
    returnChanges = {
        durability: "soft",
        returnChanges: true
    },
    busy = false;

module.exports = function(r, conn, onerror, status) {
    mq.bind("tcp://*:7000");
    return function() {
        if (!busy) {
            busy = true;
            const tz = moment().tz("America/Sao_Paulo").format("Z"),
                seconds_diff = status ? 3600 : 300,
                tables = ["events", "errors", "forms", "purchases", "mailings"],
                dependent_tables = ["traits", "pageviews", "subscriptions", "impressions", "session_goals", "visitor_goals"];

            let query = r.table("deleted_sessions").between([status, r.minval], [status, r.now().inTimezone(tz).sub(seconds_diff)], {
                index: "last_pass_created_at"
            });
            if (!status) {
                query = query.update({
                    last_pass: true
                }, returnChanges);
            } else {
                query = query.delete(returnChanges);
            }
            query = query("changes")("old_val").default([]).forEach(function(session) {
                return r.branch(session.ne(null),
                    r.table("sessions").get(session("id")).delete({
                        durability: "soft"
                    }).do(function() {
                        return r(tables).concatMap(function(table) {
                            return r.table(table).getAll(session("id"), {
                                index: "session_id"
                            }).map(function(item) {
                                return {
                                    batch: [{
                                        table: table,
                                        data: item.merge(r.branch(item.hasFields("updated_at"), {
                                            visitor_id: session("new_visitor_id"),
                                            session_id: session("new_session_id"),
                                            created_at: item("created_at").toISO8601(),
                                            ended_at: item("updated_at").toISO8601()
                                        }, {
                                            visitor_id: session("new_visitor_id"),
                                            session_id: session("new_session_id"),
                                            created_at: item("created_at").toISO8601()
                                        }))
                                    }]
                                };
                            });
                        });
                    }).do(function(items) {
                        return r(dependent_tables).forEach(function(table) {
                            return r.table(table).getAll(session("id"), {
                                index: "session_id"
                            }).delete({
                                durability: "soft",
                                returnChanges: true
                            })("changes")("old_val").default([]).map(function(item) {
                                return {
                                    batch: [{
                                        table: table,
                                        data: item.merge({
                                            id: r.literal(),
                                            created_at: item("created_at").toISO8601(),
                                            visitor_id: session("new_visitor_id"),
                                            session_id: session("new_session_id")
                                        }).merge(r.branch(table.eq("subscriptions"), {
                                            email: session("new_email")
                                        }, {})).merge(r.branch(item.hasFields("ended_at"), {
                                            ended_at: item("ended_at").toISO8601()
                                        }, r.branch(item.hasFields("updated_at"), {
                                            updated_at: item("updated_at").toISO8601()
                                        }, {})))
                                    }]
                                };
                            }).coerceTo("array").add(items);
                        });
                    }), []);
            })("batch").distinct().default([]).run().then(function(result) {
                if (result.length > 0) {
                    mq.send(JSON.stringify({
                        batch: true,
                        items: result
                    }));
                }
            }).error(onerror);
            busy = false;
        }
    };
};

```

**Comment by neumino** [comment]
Does it always fail with rethinkdbdash?

**Comment by thelinuxlich** [comment]
yes

**Comment by thelinuxlich** [comment]
rethinkdbdash 2.1.8 and node 4 btw

**Comment by neumino** [comment]
Can you provide a standalone script without mq/nano/etc. as I don't have this setup here?

**Comment by thelinuxlich** [comment]
I've updated the script

**Comment by neumino** [comment]
What version of node? Any special flag?

**Comment by thelinuxlich** [comment]
Node 4, no special flag

**Comment by neumino** [comment]
Hum, it works fine for me with node 4 (rc5), but then I guess I'm not hitting the same path as my tables are kind of scarce.

Can you set up a full stand alone script that reproduce this?

**Comment by sjmueller** [comment]
I received the same error myself, but it was due to accidentally passing a value of undefined into `filter`.  See the code example below to reproduce in a similar fashion.  FYI, I am on Node 4.1.1 and thinky 2.1.8:
```
const thinky = require('thinky')();
const type = thinky.type;
const r = thinky.r;

const Post = thinky.createModel('post_author', {
  id: type.string(),
  author_id: type.string(),
  post_id: type.string()
});
    
const author = ""; // probably should be an object
r.table('post_author')('author_id')
  .filter(author.id) // not null, but actually undefined
  .distinct().then(console.log);

// Possibly unhandled ReqlClientError: MALFORMED PROTOBUF (missing field `type`):
```

**Comment by neumino** [comment]
Thanks @sjmueller -- it's fixed in 2.1.9
Errors were not properly propagated through `distinct`
